rust/parquet/src/data_type.rs

```diff
@@ -492,8 +579,14 @@ pub(crate) mod private {
         + super::SliceAsBytes
         + PartialOrd
     {
-        /// Return the most primitive version of encoding a given type
-        fn encoded(&self) -> EncodedValue<'_>;
+        /// Encode the value directly from a higher level encoder
+        fn encode<W: std::io::Write>(values: &[Self], writer: &mut W, bit_writer: &mut BitWriter) -> Result<()>;
```

Comment:
Performance notes:

This is similar to specialisation in that it makes the required behaviours part of the trait implementation for the actual data carrying types in the code.